src/controller.cpp

```diff
@@ -75,7 +82,29 @@ void Controller::setPlan(const nav_msgs::msg::Path & path) { path_handler_.setPa
 
 void Controller::setSpeedLimit(const double & speed_limit, const bool & percentage)
 {
-  RCLCPP_ERROR(logger_, "MPPI's dynamic speed limit adjustment callback is not yet implemented!");
+  double x_vel_, y_vel_, theta_vel_;
+
+  if (speed_limit == nav2_costmap_2d::NO_SPEED_LIMIT) {
+    // Restore default value
+    x_vel_ = base_x_vel_;
+    y_vel_ = base_y_vel_;
+    theta_vel_ = base_theta_vel_;
+  } else {
+    if (percentage) {
+      // Speed limit is expressed in % from maximum speed of robot
+      x_vel_ = base_x_vel_ * speed_limit / 100.0;
+      y_vel_ = base_y_vel_ * speed_limit / 100.0;
+      theta_vel_ = base_theta_vel_ * speed_limit / 100.0;
+    } else {
+      // Speed limit is expressed in absolute value
+      double ratio = speed_limit / base_x_vel_;
+      x_vel_ = speed_limit;
+      y_vel_ = base_y_vel_ * ratio;
+      theta_vel_ = base_theta_vel_ * ratio;
+    }
+  }
+
+  optimizer_.setMaxVelocities(x_vel_, y_vel_, theta_vel_);
```

Comment:
Why delegate this to this func ? Cant we just use class fields

Reply:
Typically bad design to just grab member variables. I think good design usually calls for `set` and `get` ers